### PR TITLE
Add VOR date publication date

### PR DESCRIPTION
**Pull request by HazalCiplak**
https://github.com/elifesciences/data-hub-issues/issues/739

### CODE REVIEW

File: tests/unit_tests/docmaps/v2/codecs/elife_manuscript_test.py [hunk 1]
```diff
@@ -121,22 +124,23 @@ def test_should_not_populate_published_if_rp_publication_timestamp_is_none(self)
...
         manuscript_version_doi = get_elife_manuscript_version_doi(
-            elife_doi=DOCMAPS_QUERY_RESULT_ITEM_1['elife_doi'],
-            elife_doi_version_str=MANUSCRIPT_VERSION_1['elife_doi_version_str']
+            elife_doi=DOCMAPS_QUERY_RESULT_ITEM_WITH_VOR_VERSION['elife_doi'],
+            elife_doi_version_str=MANUSCRIPT_VOR_VERSION_1['elife_doi_version_str']
         )
         assert result == {
             'type': 'version-of-record',
             'doi': manuscript_version_doi,
+            'published': VOR_PUBLICATION_DATE_1.strftime('%Y-%m-%d'),
```
Comment (de-code) [inline tests/unit_tests/docmaps/v2/codecs/elife_manuscript_test.py]: minor: maybe use `.isoformat()` as that is what we seem to be using in other examples?

File: data_hub_api/docmaps/v2/sql/docmaps_index.sql [hunk 2]
```diff
@@ -317,6 +317,15 @@ t_rp_publication_date AS (
   WHERE rn=1
 ),
 
+t_vor_publication_date AS (
+  SELECT 
+    article_id,
+    vor_publication_date
```
Comment (de-code) [inline data_hub_api/docmaps/v2/sql/docmaps_index.sql:323]: I would expect `vor_publication_date` to be of type DATE.
I think we probably should update `v_elife_article_xml_data`.

File: data_hub_api/docmaps/v2/sql/docmaps_index.sql [hunk 3]
```diff
@@ -344,7 +353,8 @@ t_result_with_sorted_manuscript_versions_array AS (
         result.senior_editor_details,
         result.author_names_csv,
         result.evaluations,
-        CONCAT(publication.publication_date, ' ', publication.utc_publication_time) AS rp_publication_timestamp
+        CONCAT(publication.publication_date, ' ', publication.utc_publication_time) AS rp_publication_timestamp,
```
Comment (de-code) [inline data_hub_api/docmaps/v2/sql/docmaps_index.sql]: Not sure if I mentioned it in a previous PR. But here `rp_publication_timestamp` is not a timestamp but a string.
Either we name it `rp_publication_timestamp_str` and parse it in Python or parse it within BigQuery SQL.

The test data should match the type. But currently looks to be a parsed timestamp?

File: data_hub_api/docmaps/v2/codecs/elife_manuscript.py [hunk 4]
```diff
@@ -101,6 +101,11 @@ def get_docmap_elife_manuscript_output_for_vor(
     return {
         'type': 'version-of-record',
         'doi': manuscript_version_doi,
+        'published': (
+            manuscript_version['vor_publication_date'].strftime('%Y-%m-%d')
```
Comment (de-code) [inline data_hub_api/docmaps/v2/codecs/elife_manuscript.py]: minor: we could also use `.isoformat()` as we do in some other cases?

BTW how is this currently working, isn't `vor_publication_date` coming from BigQuery currently a string?

**Comment by de-code** [comment]
Not this change but I noticed the type mismatch for `rp_publication_timestamp`. I was wondering that it isn't failing. But then notice we are converting it to a `str` in this code:

```python
format_datetime_with_utc_offset(
                date_string=str(manuscript_version['rp_publication_timestamp'])
            )
```

We shouldn't use `str` to convert a timestamp to a date format. And then format it again. `rp_publication_timestamp` should already be a timestamp using the Python type `datetime`.

I don't quite understand `format_datetime_with_utc_offset`. It seems too flexible. We can pass in `date_string` but it also accepts a `datetime`. `format_datetime_with_utc_offset` should really only accept a `datetime`.

Also if the datetime already has a timezone (which it usually should), then we should use that I think.

I also would generally advise against using `dateutil` in production code. (I believe we are using it for the web api but may have come across odd behaviour)

See for example this:

```python
>>> from dateutil.parser import parse
>>> parse('1')
datetime.datetime(2023, 8, 1, 0, 0)
```

`1` clearly isn't a valid date / timestamp. But it was silently ignored.

File: data_hub_api/docmaps/v2/sql/docmaps_index.sql (hunk 2 above)
Comment (HazalCiplak) [reply data_hub_api/docmaps/v2/sql/docmaps_index.sql:323]: I have updated that view.

File: data_hub_api/docmaps/v2/codecs/elife_manuscript.py (hunk 4 above)
Comment (HazalCiplak) [reply data_hub_api/docmaps/v2/codecs/elife_manuscript.py]: it is not string it is date now maybe haven’t been updated yet in prod I have recently merged a PR to fix this

**Comment by HazalCiplak** [comment]
Updated.

File: data_hub_api/docmaps/v2/sql/docmaps_index.sql [hunk 5]
```diff
@@ -353,7 +353,13 @@ t_result_with_sorted_manuscript_versions_array AS (
         result.senior_editor_details,
         result.author_names_csv,
         result.evaluations,
-        CONCAT(publication.publication_date, ' ', publication.utc_publication_time) AS rp_publication_timestamp,
+        FORMAT_TIMESTAMP(
```
Comment (de-code) [inline data_hub_api/docmaps/v2/sql/docmaps_index.sql]: If you format it, then the field should be called `rp_publication_timestamp_str`. But why not return the timestamp to Python?

Comment (de-code) [reply data_hub_api/docmaps/v2/sql/docmaps_index.sql]: For example for the evaluations we are also using `isoformat` (assuming that the `annotation_created_timestamp` is indeed a timestamp with timezone information)

Comment (HazalCiplak) [reply data_hub_api/docmaps/v2/sql/docmaps_index.sql]: Maybe we can look into it together as I dont get what you mean.

Comment (de-code) [reply data_hub_api/docmaps/v2/sql/docmaps_index.sql]: Sure.

But `PARSE_TIMESTAMP` returns a timestamp. That is what I would expect.
`FORMAT_TIMESTAMP` then turns it back to a string. That is what I would remove. Then `rp_publication_timestamp` would indeed be a timestamp. We can use `isoformat` to format it within Python where needed. (As we already do with `annotation_created_timestamp` for example)

**Review by de-code — APPROVED** [review]

**Pull request merged**